libcxx/test/support/test_iterators.h

```diff
@@ -1083,6 +1083,23 @@ rvalue_iterator(T*) -> rvalue_iterator<T>;
 
 static_assert(std::random_access_iterator<rvalue_iterator<int*>>);
 
+// The ProxyDiffTBase allows us to conditionally specify Proxy<T>::difference_type
+// which we need in certain situations. For example when we want
+// std::weakly_incrementable<Proxy<T>> to be true.
+template <class T>
+struct ProxyDiffTBase {
```

Comment:
Question: why do we need a base class to achieve this?

Reply:
The base class allows us to have the specialization for types that have `std::iter_difference_t` in a cleaner way. I think you're right that the base class isn't strictly necessary, but we'd have a lot of duplicate code with we implemented specialization for the `Proxy` class

Reply:
As a rule of thumb, I don't mind duplication in tests so that things are known to be correct on inspection. Abstraction is fine, but only if it doesn't cast doubt on our confidence that the test is correct.

I think in this case, I'd prefer a `proxy_with_difference_type` and `proxy_missing_difference_type`. That might be deserving of its own PR, and before this one is merged so you can use it.